File: Algorithms/897_Increasing_Order_Search_Tree/Solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private final TreeNode fakeNode = new TreeNode(-1);
    private TreeNode head = fakeNode;

    public TreeNode increasingBST(TreeNode root) {
        if (root != null) {
            increasingBST(root.left);
            head.right = root;
            head = head.right;
            head.left = null;
            increasingBST(root.right);
        }
        return fakeNode.right;
    }
}

File: Algorithms/897_Increasing_Order_Search_Tree/Solution.kt
/**
 * Example:
 * var ti = TreeNode(5)
 * var v = ti.`val`
 * Definition for a binary tree node.
 * class TreeNode(var `val`: Int) {
 *     var left: TreeNode? = null
 *     var right: TreeNode? = null
 * }
 */
class Solution {
    fun increasingBST(root: TreeNode?, tail: TreeNode? = null): TreeNode? {
//        println("root = ${root?.`val`}, tail = ${tail?.`val`}")
        if (root == null) {
            return tail
        }
        val head = increasingBST(root.left, root)
        root.left = null
        root.right = increasingBST(root.right, tail)
        return head
    }
}
